releases: add tidb 4.0.10 release notes (#4608)
pingcap/docs: releases/release-4.0.10.md

        - Fix a bug that outdated metadata might cause the newly created changefeed abnormal [#1184](https://github.com/pingcap/ticdc/pull/1184)
        - Fix create receiver on closed notifier [#1199](https://github.com/pingcap/ticdc/pull/1199)
        - Fix memory accumulated when owner consume etcd update slow [#1227](https://github.com/pingcap/ticdc/pull/1227)
        - Fix the issue that `max-batch-size` does not take effect [#1253](https://github.com/pingcap/ticdc/pull/1253)
        - Fix cleanup stale tasks before capture information is constructed [#1280](https://github.com/pingcap/ticdc/pull/1280)
        - Roll back transactions to recycle db conn for MySQL sink [#1285](https://github.com/pingcap/ticdc/pull/1285)

    + Dumpling

        - Do not set `tidb_mem_quota_query` when it is not specified to avoid TiDB OOM [#233](https://github.com/pingcap/dumpling/pull/233)

    + Backup & Restore (BR)

        - Fix the issue that BR v4.0.9 cannot restore the GCS backup files of v4.0.8 [#688](https://github.com/pingcap/br/pull/688)

>         - Fix the issue that BR panics when the GCS storage URL has no prefix [#673](https://github.com/pingcap/br/pull/673)

Reviewer: ```suggestion
        - Fix the issue that BR panics when the GCS storage URL has no prefix [#673](https://github.com/pingcap/br/pull/673)
        - Disable backup stats by default to avoid BR OOM [#693](https://github.com/pingcap/br/pull/693)

```

---
title: TiDB 4.0.10 Release Notes
---

# TiDB 4.0.10 Release Notes

Release date: January 15, 2021

TiDB version: 4.0.10

## Compatibility Changes

## New Features

> 

Reviewer 2: ```suggestion

+ TiFlash

    - Add `security.redact_info_log` config, which can redact user data from logs.
    - Support lossless schema change for enum type.
    
```

## Improvements

## Bug Fixes

+ TiDB

    - Fix a concurrency bug that may cause the batch client timeout [#22336](https://github.com/pingcap/tidb/pull/22336)
    - Avoid duplicate bindings caused by concurrent baseline capture [#22295](https://github.com/pingcap/tidb/pull/22295)
    - Make baseline capture work when log level is 'debug' [#22293](https://github.com/pingcap/tidb/pull/22293)

>     - Fix gc miss locks when region merged during scanning & resolving locks [#22267](https://github.com/pingcap/tidb/pull/22267)

Reviewer 3: ```suggestion
    - Correctly GC locks when region merge happens during scanning & resolving locks [#22267](https://github.com/pingcap/tidb/pull/22267)
```

    - Return correct results for user variables of datetime type [#22143](https://github.com/pingcap/tidb/pull/22143)
    - Avoid using index merge when there are multiple table filters. [#22124](https://github.com/pingcap/tidb/pull/22124)

>     - Fixed the issue that using prepare statement to cache execution plan caused tiflash to return `Wrong precision`. [#21960](https://github.com/pingcap/tidb/pull/21960)

Reviewer 3: ```suggestion
    - Fix the `wrong precision` problem in TiFlash caused by prepare plan cache. [#21960](https://github.com/pingcap/tidb/pull/21960)
```

>     - Fix a bug that schema change will lead to incorrectness when forUpdateRead [#21596](https://github.com/pingcap/tidb/pull/21596)

Reviewer 3: ```suggestion
    - Fix the bug that schema change will lead to incorrect results. [#21596](https://github.com/pingcap/tidb/pull/21596)
```

>     - Fix a bug that doesn't add NoDefaultValueFlag when creating a table [#21474](https://github.com/pingcap/tidb/pull/21474)

Reviewer 3: ```suggestion
    - Avoid unnecessary column flag changes incurred by `ALTER TABLE`. [#21474](https://github.com/pingcap/tidb/pull/21474)
```

>     - Set dbName for hinted query block table alias [#21380](https://github.com/pingcap/tidb/pull/21380)

Reviewer 3: ```suggestion
    - Set database name for table aliases of query blocks used in optimizer hint. [#21380](https://github.com/pingcap/tidb/pull/21380)
```

    - Generate proper hint for IndexHashJoin / IndexMergeJoin [#21020](https://github.com/pingcap/tidb/pull/21020)

+ TiFlash

>     - Fix the problem that TiFlash fail to start when path_realtime_mode is true [#1322](https://github.com/pingcap/tics/pull/1322)
>     - Fixed an issue where tables created in the old version of tidb may cause the startup of the new tiflash node to fail. [#1297](https://github.com/pingcap/tics/pull/1297)

Reviewer 2: ```suggestion
    - Fixed an issue that TiFlash can't start because TiFlash failed to process the schema of TiDB with an old version.
    - Fixed an issue where TiFlash can't start due to incorrect handling of cpu_time on the RedHat system.
    - Fixed the issue that TiFlash fails to start when path_realtime_mode is true.
    - Fixed the issue of incorrect results when calling the `substr` function with three parameters.
```

## 未分类 notes。请对以下 Note 进行分类，挪动到上面 Compatibility changes、New features、Improvement 分类下

+ TiDB

    - Session: fix the duplicate binding case when updating bind info [#22368](https://github.com/pingcap/tidb/pull/22368)
    - Session: fix two cases when updating bind info [#22341](https://github.com/pingcap/tidb/pull/22341)
    - Make txn entry size limit configurable [#21843](https://github.com/pingcap/tidb/pull/21843)

+ TiKV

>     - Fix wrong mapping between readies and peers by offset [#9409](https://github.com/tikv/tikv/pull/9409)
>     - Fix some logs not being redact when `security.redact-info-log = true` [#9314](https://github.com/tikv/tikv/pull/9314)

Reviewer 4: ```suggestion
    - Fix wrong mapping between ready and peer [#9409](https://github.com/tikv/tikv/pull/9409)
    - Fix some logs are not redacted when `security.redact-info-log` is on [#9314](https://github.com/tikv/tikv/pull/9314)
```

> + PD

Reviewer 5: some changes:
# Bug Fixes
- Fix the id allocation is not monotonic [#3308](https://github.com/tikv/pd/pull/3308) [#3323](https://github.com/tikv/pd/pull/3323)
- Fix the pd client could be blocked in some cases [#3285](https://github.com/pingcap/pd/pull/3285)
# New features
- Support configuration `enable-redact-log` to enable log desensitization or not. [#3266](https://github.com/pingcap/pd/pull/3266)
# Improvement
- Optimize the store state filter metrics [#3100](https://github.com/tikv/pd/pull/3100)
- Upgrade go.etcd.io/bbolt to v1.3.5 [#3331](https://github.com/tikv/pd/pull/3331)

    - Fix the pd client could be blocked in some cases [#3285](https://github.com/pingcap/pd/pull/3285)
    - Support configuration `enable-redact-log` to enable log desensitization or not. [#3266](https://github.com/pingcap/pd/pull/3266)

> + TiFlash
> 
>     - Add `security.redact_info_log` config, which redacts user data from logs [#1333](https://github.com/pingcap/tics/pull/1333)
> 

Reviewer 2: can be removed.
```suggestion
```

---
title: TiDB 4.0.10 Release Notes
---

# TiDB 4.0.10 Release Notes

Release date: January 15, 2021

TiDB version: 4.0.10

## Compatibility Changes

## New Features

+ PD

    - Support configuration `enable-redact-log` to enable log desensitization or not [#3266](https://github.com/pingcap/pd/pull/3266)

+ TiFlash

    - Add `security.redact_info_log` config, which can redact user data from logs

>     - Support lossless schema change for enum type

Reviewer 2: remove this line.
```suggestion
```

## Improvements

+ TiDB

    - Make txn entry size limit configurable [#21843](https://github.com/pingcap/tidb/pull/21843)

> + PD

Reviewer 6: ```suggestion
+ TiKV
```

    - Set database name for table aliases of query blocks used in optimizer hint. [#21380](https://github.com/pingcap/tidb/pull/21380)
    - Generate proper hint for IndexHashJoin / IndexMergeJoin [#21020](https://github.com/pingcap/tidb/pull/21020)

+ TiKV

    - Fix wrong mapping between ready and peer [#9409](https://github.com/tikv/tikv/pull/9409)
    - Fix some logs are not redacted when `security.redact-info-log` is on [#9314](https://github.com/tikv/tikv/pull/9314)

+ PD

    - Fix the id allocation is not monotonic [#3308](https://github.com/tikv/pd/pull/3308) [#3323](https://github.com/tikv/pd/pull/3323)
    - Fix the pd client could be blocked in some cases [#3285](https://github.com/pingcap/pd/pull/3285)

+ TiFlash

    - Fixed an issue that TiFlash can't start because TiFlash failed to process the schema of TiDB with an old version
    - Fixed an issue where TiFlash can't start due to incorrect handling of cpu_time on the RedHat system
    - Fixed the issue that TiFlash fails to start when path_realtime_mode is true
    - Fixed the issue of incorrect results when calling the `substr` function with three parameters

> 

Reviewer 2: ```suggestion
Fixed the issue that TiFlash does not support Enum type change even if the change is lossless.
```

## 未分类 notes。请对以下 Note 进行分类，挪动到上面 Compatibility changes、New features、Improvement 分类下

> + Tools
> 
>     - TiCDC
> 
>         * Fix a bug that MySQL connection is not recycled when some transactions executed with failure [#1285](https://github.com/pingcap/ticdc/pull/1285)
>         * Fixed bug in passing `max-batch-size` to codec [#1253](https://github.com/pingcap/ticdc/pull/1253)
>         * Enable Unified Sorter by default [#1230](https://github.com/pingcap/ticdc/pull/1230)
>         * Fix a bug that the cdc owner might consume too much memory in the etcd watch client [#1227](https://github.com/pingcap/ticdc/pull/1227)
>         * Fix a bug that outdated metadata could cause the newly created changefeed abnormal. [#1184](https://github.com/pingcap/ticdc/pull/1184)

Reviewer: ```suggestion
# TiCDC

## Improvements

* Enable old value if protocol is maxwell [#1144](https://github.com/pingcap/ticdc/pull/1144)
* Enable Unified Sorter by default [#1230](https://github.com/pingcap/ticdc/pull/1230)

## Bug Fixes

* Fix maxwell issues, including base64 data output and output tso to unix timestamp [#1173](https://github.com/pingcap/ticdc/pull/1173)
* Fix a bug that outdated metadata could cause the newly created changefeed abnormal [#1184](https://github.com/pingcap/ticdc/pull/1184)
* Fix create receiver on closed notifier [#1199](https://github.com/pingcap/ticdc/pull/1199)
* Fix memory accumulated when owner consume etcd update slow [#1227](https://github.com/pingcap/ticdc/pull/1227)
* Fixed `max-batch-size` does not take effect [#1253](https://github.com/pingcap/ticdc/pull/1253)
* Fix cleanup stale tasks before capture info is constructed [#1280](https://github.com/pingcap/ticdc/pull/1280)
* Rollback txn to recycle db conn for MySQL sink [#1285](https://github.com/pingcap/ticdc/pull/1285)

# TiDB-binlog

## Bug Fixes

* Fix the issue that when amend txn enabled. Drainer chooses the wrong schema version of DDL to generate the wrong SQL [#1033](https://github.com/pingcap/tidb-binlog/pull/1033)

# BR

## Bug Fixes

* Fix the issue that br 4.0.9 cannot restore br 4.0.8 gcs backup files [#688](https://github.com/pingcap/br/pull/688)
* Fix the issue that br paniced when gcs storage url has not prefix [#673](https://github.com/pingcap/br/pull/673)

# Dumpling

## Improvements

* Support check un-recognized arguments and print current progress during dumping
 [#228](https://github.com/pingcap/dumpling/pull/228)

## Bug Fixes

* Don't set tidb_mem_quota_query when it's not specified to avoid TiDB OOM
 [#233](https://github.com/pingcap/dumpling/pull/233)

# TiDB-Lightning

## Improvements

* Support retry S3 read on error [#533](https://github.com/pingcap/tidb-lightning/pull/533)

## Bug Fixes

* Fix a bug that the retry scan region key is not encoded, which causes the retryed region is not split [#531](https://github.com/pingcap/tidb-lightning/pull/531)
* Fix a bug that lost create table error during multiple tables created [#530](https://github.com/pingcap/tidb-lightning/pull/530)
* Fix the bug that lightning will trim all the empty sep when trim-last-sep is true, which causes "column count mismatch" in tidb backend [#535](https://github.com/pingcap/tidb-lightning/pull/535)

```
